### PR TITLE
feat(PLM-387): Support conditional existence check field

### PR DESCRIPTION
**Pull request by kilikkuo**
[Why]
There is a need to support more complicated conditional fields existenc check

[How]
1. Add COND_EXIST and its validator
2. Add deprecating message for old checks
3. Add test cases

[PLM-387]

Reason
--------------

Changes
--------------

Test Scope
--------------

Checks
--------------
- [x] Unit tests are included or not applicable.

[PLM-387]: https://hardcoretech.atlassian.net/browse/PLM-387?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: data_spec_validator/spec/defines.py
Comment (raviwugo) [inline data_spec_validator/spec/defines.py:204]: Is there better way to manage the deprecation checks? Recalled [last PR](https://github.com/hardcoretech/data-spec-validator/pull/16/files#diff-1276fe99388885c634523e19da8aaad82fe11b675288102d5ca3e3e56b0f3872R197) there's also a deprecated `extra` keyword arg (at above line 201).  The purpose of the question is to better track how many deprecation items we have and plan the actual removal for these items. (Or maybe this is a non-issue if all those deprecation we actually keep in same interface)

File: data_spec_validator/spec/validators.py
Comment (raviwugo) [inline data_spec_validator/spec/validators.py:477]: Do we want to have a define for these `ALLOW_UNKNOWN`, `ANY`, `WITH`, `WITHOUT`, etc keywords?

File: data_spec_validator/spec/validators.py
Comment (raviwugo) [inline data_spec_validator/spec/validators.py:480]: Shall the error message be like `not a listed value`?

File: data_spec_validator/spec/validators.py
Comment (raviwugo) [inline data_spec_validator/spec/validators.py]: Maybe have the msg like `Any of the {", ".join(any_keys)} must exist`

File: data_spec_validator/spec/validators.py
Comment (raviwugo) [inline data_spec_validator/spec/validators.py:82]: The `optional` here for `COND_EXIST` added another layer than the ordinary `optional`:

`Checker([COND_EXIST], optional=True, COND_EXIST=dict(WITH=['a'], WITHOUT=['c']))` means the named key can be optional AND actual data can contain unknown field value.

But previous usage `Checker([STR], optional=True)` simply means the assigned `field` is optional.

Do we need to separate the usage of the keyword arg?

File: CHANGELOG.md
Comment (cmj0121) [inline CHANGELOG.md]: good

**Review by cmj0121 — APPROVED** [review]

File: data_spec_validator/spec/validators.py
Comment (raviwugo) [inline data_spec_validator/spec/validators.py]: Why need to have `isinstance(value, UnknownFieldValue)` here for the `ANY` condition check?

File: test/test_spec.py
Comment (raviwugo) [inline test/test_spec.py]: I think from the interface, the expected behavior is failing the `dict(a=1)` (not have ANY[b] in field `a` checker) and `dict(b=1)` (not have ANY[a] in field `b` checker).

But since the checks have condition `if any_keys and isinstance(value, UnknownFieldValue)` added, so if there's no `dict(d=1)` passed in the data the ANY checker will be by-passed.

File: data_spec_validator/spec/defines.py
Comment (kilikkuo) [reply data_spec_validator/spec/defines.py:204]: I ... don't have any thoughts for now.

What I can think of is to make the checks a CustomEnum. e.g.
```
class SomeEnum:
    INT = 'int', 'stable'
    BOOL = 'bool', 'stable'
    JSON = 'json', 'deprecating'
```
Then the supporting status of each check is much clear.

`extra` is a keyword for the `__init__` constructor, I suppose it can live longer for capability.
Any thoughts ?

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py:477]: I don't want to make them constant for now because I'm still trying to give them a better name.

Instead of (`WITH`, `WITHOUT`, `ANY`) ... do you think (`COEXIST`, `EXCLUDE`, `ANY_OF`) is better?

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py:480]: If the validator runs into here for a certain field and it's not configured `optional=True` (which means `allow_unknown` will be False), then the error message should indicate that the field must exist.
So there's an error wrapper here [1] to rephrase the message.  So actually we don't really use this `must exist` message, it's a visual hint for the developer.

[1] https://github.com/hardcoretech/data-spec-validator/blob/develop/data_spec_validator/spec/actions.py#L8-L9

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py]: I think we can have a `UnknowFieldValueError` which contains an informative message that will be concatenated in this error message [1]

[1] https://github.com/hardcoretech/data-spec-validator/blob/2fcbdc1b754bcbd7dd26e9b73c631cd962133956/data_spec_validator/spec/actions.py#L9

will address

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py:82]: Yeah, the original design of `optional` in `Checker([STR], optional=True)` is to support the absence of the named `field` key.

In `def __validate_field`, the target field value is extracted and `UnknownFieldValue` is used to represent the value (in order to distinguish it from `None`). Then inside each validator, we can focus on the `value validation`. That is the reason I convert the `optional` to `allow_unknown`.

In short, `optional` is the client-facing keyword, the `ALLOW_UNKNOWN ` is used internally.
So we should interpret the following statement in this way,
```
a = Checker([COND_EXIST], optional=True, COND_EXIST=dict(WITH=['b'], WITHOUT=['c']))
```
Field `a` can be optional, but if it appears, then COND_EXIST comes to play (`b` must exist, `c` must not)

Any comment would be appreciated, just trying to make the interface as intuitive as possible

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py]: Hmm, indeed, seems that we don't need it.
Let' me try to remove it

File: data_spec_validator/spec/defines.py
Comment (kilikkuo) [reply data_spec_validator/spec/defines.py:204]: follow-up,  https://hardcoretech.atlassian.net/browse/PLM-422

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py:477]: ANY is not the scope of COND_EXIST now

File: data_spec_validator/spec/validators.py
Comment (kilikkuo) [reply data_spec_validator/spec/validators.py]: removed

**Comment by kilikkuo** [comment]
@raviwugo , please have a second review regarding the changes we discussed, thanks

File: README.md
Comment (raviwugo) [inline README.md]: Maybe show the example of multiple set combination and explain the AND validation logic instead of a single set in README?

**Review by raviwugo — APPROVED** [review]
👏 👏 👏 👏 👏 👏 👏 👏  棒棒噠

**Pull request merged**
